File: python/src/altengine/_http.py
"""Minimal JSON-over-HTTP transports (sync + async) shared by all service clients."""

from __future__ import annotations

import os
import random
import time
from typing import Any, Dict, Mapping, Optional

import httpx

from .errors import AltEngineError, AltEngineNetworkError

#: Production API origin used when no override is given.
DEFAULT_BASE_URL = "https://api.altengine.net"
#: Where ``altengine dev`` (the local emulator) listens by default.
DEV_BASE_URL = "http://127.0.0.1:9191"

_DEFAULT_MAX_ATTEMPTS = 3
_DEFAULT_BASE_DELAY = 0.25
_DEFAULT_MAX_DELAY = 4.0


def resolve_base_url(base_url: Optional[str], dev: bool) -> str:
    """Resolution order: explicit ``base_url`` → ``dev`` → ``ALTENGINE_URL`` env
    var → production."""
    if base_url:
        return base_url.rstrip("/")
    if dev:
        return DEV_BASE_URL
    return (os.environ.get("ALTENGINE_URL") or DEFAULT_BASE_URL).rstrip("/")


def _clean_query(query: Optional[Mapping[str, Any]]) -> Optional[Dict[str, str]]:
    if not query:
        return None
    out = {k: str(v).lower() if isinstance(v, bool) else str(v) for k, v in query.items() if v is not None}
    return out or None


def _to_api_error(res: httpx.Response) -> AltEngineError:
    code, message, details = "INTERNAL", f"HTTP {res.status_code}", None
    try:
        parsed = res.json()
        err = parsed.get("error") if isinstance(parsed, dict) else None
        if isinstance(err, dict):
            code = err.get("code") or code
            message = err.get("message") or message
            details = err.get("details")
    except Exception:
        pass  # non-JSON error body; keep the status-derived defaults
    retry_after = None
    ra = res.headers.get("retry-after")
    if ra is not None:
        try:
            retry_after = float(ra)
        except ValueError:
            pass
    return AltEngineError(code, message, res.status_code, details, retry_after)


def _backoff(attempt: int, err: Optional[AltEngineError]) -> float:
    delay = min(_DEFAULT_BASE_DELAY * 2 ** (attempt - 1), _DEFAULT_MAX_DELAY)
    if err is not None and err.retry_after is not None:
        delay = max(delay, err.retry_after)
    # Full jitter keeps concurrent retries from stampeding in unison.
    return delay * (0.5 + random.random() * 0.5)


class Http:
    """Synchronous transport over ``httpx.Client``."""

    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: Optional[str] = None,
        dev: bool = False,
        timeout: float = 30.0,
        max_attempts: int = _DEFAULT_MAX_ATTEMPTS,
        client: Optional[httpx.Client] = None,
    ) -> None:
        self.base_url = resolve_base_url(base_url, dev)
        self.api_key = api_key if api_key is not None else os.environ.get("ALTENGINE_API_KEY")
        self.max_attempts = max_attempts
        self._client = client or httpx.Client(timeout=timeout)

    def close(self) -> None:
        self._client.close()

    def request(
        self,
        method: str,
        path: str,
        *,
        query: Optional[Mapping[str, Any]] = None,
        body: Any = None,
        headers: Optional[Mapping[str, str]] = None,
        retry: bool = True,
    ) -> Any:
        attempts = self.max_attempts if retry else 1
        attempt = 0
        while True:
            attempt += 1
            try:
                return self._once(method, path, query, body, headers)
            except AltEngineError as err:
                if not err.retryable or attempt >= attempts:
                    raise
                time.sleep(_backoff(attempt, err))
            except AltEngineNetworkError:
                if attempt >= attempts:
                    raise
                time.sleep(_backoff(attempt, None))

    def _once(self, method, path, query, body, headers) -> Any:
        hdrs = dict(headers or {})
        if self.api_key:
            hdrs["authorization"] = f"Bearer {self.api_key}"
        try:
            res = self._client.request(
                method,
                self.base_url + path,
                params=_clean_query(query),
                json=body,
                headers=hdrs,
            )
        except httpx.HTTPError as exc:
            raise AltEngineNetworkError(f"request failed: {method} {path}: {exc}") from exc
        if res.status_code < 200 or res.status_code > 299:
            raise _to_api_error(res)
        if res.status_code == 204 or not res.content:
            return None
        return res.json()


class AsyncHttp:
    """Asynchronous transport over ``httpx.AsyncClient``."""

    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: Optional[str] = None,
        dev: bool = False,
        timeout: float = 30.0,
        max_attempts: int = _DEFAULT_MAX_ATTEMPTS,
        client: Optional[httpx.AsyncClient] = None,
    ) -> None:
        self.base_url = resolve_base_url(base_url, dev)
        self.api_key = api_key if api_key is not None else os.environ.get("ALTENGINE_API_KEY")
        self.max_attempts = max_attempts
        self._client = client or httpx.AsyncClient(timeout=timeout)

    async def close(self) -> None:
        await self._client.aclose()

    async def request(
        self,
        method: str,
        path: str,
        *,
        query: Optional[Mapping[str, Any]] = None,
        body: Any = None,
        headers: Optional[Mapping[str, str]] = None,
        retry: bool = True,
    ) -> Any:
        import asyncio

        attempts = self.max_attempts if retry else 1
        attempt = 0
        while True:
            attempt += 1
            try:
                return await self._once(method, path, query, body, headers)
            except AltEngineError as err:
                if not err.retryable or attempt >= attempts:
                    raise
                await asyncio.sleep(_backoff(attempt, err))
            except AltEngineNetworkError:
                if attempt >= attempts:
                    raise
                await asyncio.sleep(_backoff(attempt, None))

    async def _once(self, method, path, query, body, headers) -> Any:
        hdrs = dict(headers or {})
        if self.api_key:
            hdrs["authorization"] = f"Bearer {self.api_key}"
        try:
            res = await self._client.request(
                method,
                self.base_url + path,
                params=_clean_query(query),
                json=body,
                headers=hdrs,
            )
        except httpx.HTTPError as exc:
            raise AltEngineNetworkError(f"request failed: {method} {path}: {exc}") from exc
        if res.status_code < 200 or res.status_code > 299:
            raise _to_api_error(res)
        if res.status_code == 204 or not res.content:
            return None
        return res.json()


def seg(value: Any) -> str:
    """Encode one path segment (instance/index/collection/key names)."""
    from urllib.parse import quote

    return quote(str(value), safe="")

File: php/src/Http.php
<?php

declare(strict_types=1);

namespace AltEngine;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Exception\GuzzleException;

/**
 * Minimal JSON-over-HTTP transport shared by all service clients.
 *
 * @internal
 */
final class Http
{
    /** Production API origin used when no override is given. */
    public const DEFAULT_BASE_URL = 'https://api.altengine.net';
    /** Where `altengine dev` (the local emulator) listens by default. */
    public const DEV_BASE_URL = 'http://127.0.0.1:9191';

    private const MAX_ATTEMPTS = 3;
    private const BASE_DELAY = 0.25;
    private const MAX_DELAY = 4.0;

    private string $baseUrl;
    private ?string $apiKey;
    private GuzzleClient $client;
    private int $maxAttempts;

    /**
     * @param array{api_key?: string, base_url?: string, dev?: bool, timeout?: float,
     *              max_attempts?: int, client?: GuzzleClient} $options
     */
    public function __construct(array $options = [])
    {
        $baseUrl = $options['base_url']
            ?? (($options['dev'] ?? false) ? self::DEV_BASE_URL : null)
            ?? (getenv('ALTENGINE_URL') ?: null)
            ?? self::DEFAULT_BASE_URL;
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->apiKey = $options['api_key'] ?? (getenv('ALTENGINE_API_KEY') ?: null);
        $this->maxAttempts = $options['max_attempts'] ?? self::MAX_ATTEMPTS;
        $this->client = $options['client'] ?? new GuzzleClient([
            'timeout' => $options['timeout'] ?? 30.0,
            'http_errors' => false,
        ]);
    }

    /**
     * @param array<string, mixed>|null $query
     * @param array<string, string>|null $headers
     */
    public function request(
        string $method,
        string $path,
        ?array $query = null,
        mixed $body = null,
        ?array $headers = null,
        bool $retry = true,
    ): mixed {
        $attempts = $retry ? $this->maxAttempts : 1;
        $attempt = 0;
        while (true) {
            $attempt++;
            try {
                return $this->once($method, $path, $query, $body, $headers);
            } catch (AltEngineError $err) {
                if (!$err->retryable() || $attempt >= $attempts) {
                    throw $err;
                }
                $this->backoff($attempt, $err->retryAfter);
            } catch (AltEngineNetworkError $err) {
                if ($attempt >= $attempts) {
                    throw $err;
                }
                $this->backoff($attempt, null);
            }
        }
    }

    private function backoff(int $attempt, ?float $retryAfter): void
    {
        $delay = min(self::BASE_DELAY * (2 ** ($attempt - 1)), self::MAX_DELAY);
        if ($retryAfter !== null && $retryAfter > $delay) {
            $delay = $retryAfter;
        }
        // Full jitter keeps concurrent retries from stampeding in unison.
        usleep((int) ($delay * (0.5 + mt_rand() / mt_getrandmax() * 0.5) * 1_000_000));
    }

    /**
     * @param array<string, mixed>|null $query
     * @param array<string, string>|null $headers
     */
    private function once(string $method, string $path, ?array $query, mixed $body, ?array $headers): mixed
    {
        $hdrs = $headers ?? [];
        if ($this->apiKey !== null) {
            $hdrs['authorization'] = 'Bearer ' . $this->apiKey;
        }
        $options = ['headers' => $hdrs];
        if ($query !== null) {
            $options['query'] = array_map(
                static fn ($v) => is_bool($v) ? ($v ? 'true' : 'false') : (string) $v,
                array_filter($query, static fn ($v) => $v !== null),
            );
        }
        if ($body !== null) {
            $options['json'] = $body;
        }

        try {
            $res = $this->client->request($method, $this->baseUrl . $path, $options);
        } catch (GuzzleException $exc) {
            throw new AltEngineNetworkError("request failed: {$method} {$path}: {$exc->getMessage()}", 0, $exc);
        }

        $status = $res->getStatusCode();
        $raw = (string) $res->getBody();
        if ($status < 200 || $status > 299) {
            $code = 'INTERNAL';
            $message = "HTTP {$status}";
            $details = null;
            $parsed = json_decode($raw, true);
            if (is_array($parsed) && isset($parsed['error']) && is_array($parsed['error'])) {
                $code = $parsed['error']['code'] ?? $code;
                $message = $parsed['error']['message'] ?? $message;
                $details = $parsed['error']['details'] ?? null;
            }
            $ra = $res->getHeaderLine('retry-after');
            $retryAfter = is_numeric($ra) ? (float) $ra : null;
            throw new AltEngineError($code, $message, $status, $details, $retryAfter);
        }
        if ($status === 204 || $raw === '') {
            return null;
        }
        return json_decode($raw, true);
    }

    /** Encode one path segment (instance/index/collection/key names). */
    public static function seg(string|int $value): string
    {
        return rawurlencode((string) $value);
    }

    /**
     * Encode a namespace path segment. The empty (default) namespace can't ride
     * a URL path — routers collapse the resulting "//" — so it travels as the
     * reserved sentinel `_default` (the server maps it back to "").
     */
    public static function nsSeg(string $namespace): string
    {
        return rawurlencode($namespace === '' ? '_default' : $namespace);
    }
}
